include/boost/geometry/algorithms/detail/point_is_spike_or_equal.hpp

```diff
@@ -42,33 +65,29 @@ namespace detail
 // So specify last point first, then (a,b)
 // The segment's orientation does matter: if lp is to the right of b
 // no spike is reported
-template <typename Point1, typename Point2, typename Point3>
-static inline bool point_is_spike_or_equal(Point1 const& last_point,
-            Point2 const& segment_a,
-            Point3 const& segment_b)
+template
+<
+    typename Point1, typename Point2, typename Point3,
+    typename SideStrategy
+>
+static inline bool point_is_spike_or_equal(Point1 const& last_point, // prev | back
+                                           Point2 const& segment_a,  // next | back - 2
+                                           Point3 const& segment_b,  // curr | back - 1 | spike's vertex
+                                           SideStrategy const& strategy)
 {
-    typedef typename strategy::side::services::default_strategy
-    <
-        typename cs_tag<Point1>::type
-    >::type side_strategy;
-
-    int const side = side_strategy::apply(last_point, segment_a, segment_b);
+    int const side = strategy.apply(segment_a, segment_b, last_point);
     if (side == 0)
     {
         // Last point is collinear w.r.t previous segment.
-        // Check if it is equal
-        int const sgn_x1 = sign_of_difference<0>(last_point, segment_b);
-        int const sgn_y1 = sign_of_difference<1>(last_point, segment_b);
-        if (sgn_x1 == 0 && sgn_y1 == 0)
-        {
-            return true;
-        }
-
-        // Check if it moves forward
-        int const sgn_x2 = sign_of_difference<0>(segment_b, segment_a);
-        int const sgn_y2 = sign_of_difference<1>(segment_b, segment_a);
-
-        return sgn_x1 != sgn_x2 || sgn_y1 != sgn_y2;
+#ifdef BOOST_GEOMETRY_ENABLE_POINT_IS_SPIKE_OR_EQUAL_TEST
+        bool r1 = collinear_point_is_spike_or_equal(last_point, segment_a, segment_b);
+        bool r2 = direction_code(segment_a, segment_b, last_point) < 1;
+        if (r1 != r2)
+            std::cout << "spike detection failure with: " << r1 << " " << r2 << std::endl;
```

Comment:
Temporarily I added code comparing the previous and current version of spike detection enabled when defined `BOOST_GEOMETRY_ENABLE_POINT_IS_SPIKE_OR_EQUAL_TEST`. In case you wanted to check what's the difference, i.e. where the result was wrong. I noticed several differences in some buffer tests (AFAIR multi_polygon), union and more.